Sources/XCTest/XCTestCaseType.swift

```diff
@@ -0,0 +1,119 @@
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//
+//  XCTestCaseType.swift
+//  Base protocol (and extension with default methods) for test cases
+//
+
+public typealias TestMethod = XCTestCaseType throws -> Void
+
+public protocol XCTestCaseType {
+    init()
+
+    func setUp()
+    func tearDown()
+
+    // In the Objective-C version of XCTest, it is possible to discover all tests when the test is executed by asking the runtime for all methods and looking for the string "test". In Swift, we ask test providers to tell us the list of tests by implementing this property.
+    static var allTests: [(String, TestMethod)] { get }
+}
+
+/// Wrapper function allowing a static reference to a test method to fit 
+/// the signature required by XCTestCaseType.allTests
+public func test<T: XCTestCaseType>(testFunc: T -> () throws -> Void) -> TestMethod {
+    return { testCaseType in
+        guard let testCase: T = testCaseType as? T else {
+            fatalError("Attempt to invoke test on class \(T.self) with incompatible instance type \(testCaseType.dynamicType)")
+        }
+
+        try testFunc(testCase)()
+    }
+}
+
+extension XCTestCaseType {
+
+    public var continueAfterFailure: Bool {
+        get {
+            return true
+        }
+        set {
+            // TODO: When using the Objective-C runtime, XCTest is able to throw an exception from an assert and then catch it at the frame above the test method. This enables the framework to effectively stop all execution in the current test. There is no such facility in Swift. Until we figure out how to get a compatible behavior, we have decided to hard-code the value of 'true' for continue after failure.
+        }
+    }
+
+    public static func invokeTest() {
+        let tests = self.allTests
+        var totalDuration = 0.0
+        var totalFailures = 0
+        var unexpectedFailures = 0
+        let overallDuration = measureTimeExecutingBlock {
+            for (name, test) in tests {
```

Comment:
The `test` variable here shadows the `test()` function above.

At first I wondered whether the `test()` function should be renamed to something less likely to clash, but I actually am quite fond of the name--it's simple and easy for test writers to understand.

Instead, how about renaming this variable? Perhaps `testClosure` or `wrappedTest` or something?